File: A06 Nodes Related/002 Binary Tree/Leetcode 103 Binary Zigzag Level Order Traversal/Leetcode 103.java
/*


*/ 

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        //corner case
        if (root == null) return res; 
        
        Queue<TreeNode> queue = new LinkedList<>();  
        queue.add(root);
        int level = 0;          //addon modification base on level order traversal
        while(!queue.isEmpty()){
            List<Integer> tmp = new ArrayList<>();
            int size = queue.size();
            for(int i = 0; i < size; i++){
                TreeNode cur = queue.remove();
                tmp.add(cur.val);
                if (cur.left != null){
                    queue.add(cur.left); 
                }
                if (cur.right != null){
                    queue.add(cur.right);
                }
            }
            if (level % 2 ==1) {         //addon modification base on level order traversal
                Collections.reverse(tmp);
            }
            res.add(tmp);
            level ++;       //addon modification base on level order traversal
        }
        return res; 
    }
}

File: A06 Nodes Related/002 Binary Tree/Leetcode 103 Binary Zigzag Level Order Traversal/Leetcode 103.py
###########
Complexity: time complexity is O(n), where n is number of nodes in our binary tree. Space complexity is also O(n), because our result has this size in the end. If we do not count output as additional space, then it will be O(w), where w is width of tree. It can be reduces to O(1) I think if we traverse levels in different order directly, but it is just not worth it.

#############
class Solution:
    def zigzagLevelOrder(self, root):
        if not root: return []
        queue = deque([root])
        result, direction = [], 1
        
        while queue:
            level = []
            for i in range(len(queue)):
                node = queue.popleft()
                level.append(node.val)
                if node.left:  queue.append(node.left)
                if node.right: queue.append(node.right)
            result.append(level[::direction])
            direction *= (-1)
        return result